### PR TITLE
🏗 Replace `gulp-closure-compiler` with `google-closure-compiler` (reprise)

### LINKED ISSUE #21839: gulp dist fails on Windows
## What's the issue?

**gulp dist fails on Windows because it fails to create sub directories**

The issue is that 'gulp-closure-compiler' uses 'graceful-fs' which is not capable of recursively create subdirectories in Windows
I was able to workaround the issue by manually editing **node_modules/gulp-closure-compiler.js**. 

**Code sample**
File: node_modules/gulp-closure-compiler.js
`    // Create directory for output file if it doesn't exist.
    if (opt.fileName && !fs.existsSync(path.dirname(opt.fileName))) {
      fs.mkdirSync(path.dirname(opt.fileName));
    }`

I modified the code to use mkdirp directly:
`    // Create directory for output file if it doesn't exist.
    if (opt.fileName && !fs.existsSync(path.dirname(opt.fileName))) {
      mkdirp(path.dirname(opt.fileName));
    }`

**Error message**
fs.js:1**14
    throw err;
    ^

Error: ENOENT: no such file or directory, mkdir 'build/cc/ads\inabox'
    at Object.mkdirSync (fs.js:752:3)
    at Stream.endStream (c:\src\amphtml\node_modules\gulp-closure-compiler\index.js:110:10)
    at _end (c:\src\amphtml\node_modules\through\index.js:65:9)
    at Stream.stream.end (c:\src\amphtml\node_modules\through\index.js:74:5)
    at DestroyableTransform.onend (c:\src\amphtml\node_modules\vinyl-fs\node_modules\readable-stream\lib\_stream_readable.js:523:10)
    at Object.onceWrapper (events.js:277:13)
    at DestroyableTransform.emit (events.js:194:15)
    at c:\src\amphtml\node_modules\vinyl-fs\node_modules\readable-stream\lib\_stream_readable.js:965:16
    at process._tickCallback (internal/process/next_tick.js:61:11)**

## How do we reproduce the issue?

1 - Setup a Windows machine following: https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md
2 - Build: gulp dist

## What browsers are affected?
N/A

## Which AMP version is affected?
master

##

### PR DESCRIPTION
**Pull request by rsimha**
This PR replaces `gulp-closure-compiler` (which has not received an update in 3 years) with the native `gulp` plugin that is part of `google-closure-compiler`. With this, we can do custom text replacements on `gulp.src` and pass them directly to closure compiler without having to copy the intermediate output to a temp directory.

This is the same as #21868, which had to be reverted due to the following bugs:
- To speed up closure compiler, we must use `+TieredCompilation` and not `-TieredCompilation`
- The calls to `replace` and `shortenLicense` must be done once per output file instead of once per input file (this will be further addressed by #21832)

Addresses https://github.com/ampproject/amphtml/issues/21828#issuecomment-483487677.
Partial fix for #21828
Follow up to #21868
Follow up to #21830
Follow up to #16799
Fixes #21839

### CODE REVIEW

**Comment by rsimha** [comment]
Use this link to hide whitespace-only changes: https://github.com/ampproject/amphtml/pull/21900/files?w=1

**Comment by rsimha** [comment]
I've verified that the running time for `gulp check-types` and `gulp dist --fortesting --noextensions` [on Travis](https://travis-ci.org/ampproject/amphtml/jobs/521604271) is now back to where it was.

This is ready for review.

File: build-system/tasks/compile.js [hunk 1]
```diff
@@ -353,98 +346,99 @@ function compile(entryModuleFilenames, outputDir, outputFilename, options) {
...
+      // Turn off warning for "Unknown @define" since we use define to pass
+      // args such as FORTESTING to our runner.
+      jscomp_off: ['unknownDefines'],
+      define,
+      hide_warnings_for: hideWarningsFor,
     };
 
     // For now do type check separately
-    if (argv.typecheck_only || checkTypes) {
+    if (options.checkTypes || options.typeCheckOnly) {
```
Comment (dvoytenko) [inline build-system/tasks/compile.js]: Is `argv` option not used anywhere?

Comment (rsimha) [reply build-system/tasks/compile.js]: I searched through the code, and it isn't documented or used anywhere. As of today, `gulp dist --typecheck_only` has been replaced by `gulp check-types`.

**Pull request merged**